### PR TITLE
ENH: stats: add Zipfian (different from Zipf/zeta) distribution

### LINKED ISSUE #11523: scipy.stats.zipf doesn't implement zipf distribution
Scipy implements the zeta distribution for scipy.stats.zipf instead of the zipf distribution.  There is no such thing as scipy.stats.zeta, so this affects users who wish to utilize either distribution.  The current scipy.stats.zipf should be renamed to scipy.stats.zeta and the zipf function should be implemented instead.  

#### Reproducing code example:
From `help(scipy.stats.norm.zipf)` the PMF described by the zipf documentation is the PMF for the Zeta distribution, not the Zipf distribution.  
```
class zipf_gen(scipy.stats._distn_infrastructure.rv_discrete)
 |  A Zipf discrete random variable.
 |  
 |  %(before_notes)s
 |  
 |  Notes
 |  -----
 |  The probability mass function for `zipf` is:
 |  
 |  .. math::
 |  
 |      f(k, a) = \frac{1}{\zeta(a) k^a}
 |  
 |  for :math:`k \ge 1`.
 |  
 |  `zipf` takes :math:`a` as shape parameter. :math:`\zeta` is the
 |  Riemann zeta function (`scipy.special.zeta`)
```

#### More Details
See the Wikipedia page for the zipf and zeta distributions.  Wikipedia calls it out as "Zipf distribution is related to the zeta distribution, but is not identical."
https://en.wikipedia.org/wiki/Zipf%27s_law
https://en.wikipedia.org/wiki/Zeta_distribution

#### Scipy/Numpy/Python version information:
```
>>> import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
1.4.1 1.18.1 sys.version_info(major=3, minor=6, micro=9, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-11523

#### What does this implement/fix?
gh-11523 notes that `scipy.stats.zipf` implements the zeta distribution. The zeta distribution is sometime times called the Zipf distribution, but Zipf/Zipfian is actually the name of distinct (more general) distribution.  This PR adds the Zipfian distribution and clarifies the difference between the two in the documentation. As an additional step to address the issue, we might want to add `zeta` as an alias for the existing `zipf` and possibly doc-deprecate `zipf`.

#### Additional information
I've tried to follow the [guide](http://scipy.github.io/devdocs/dev/contributor/adding_new.html#adding-new) to adding a new distribution, but please let me know what I need to change.

### CODE REVIEW

File: doc/source/tutorial/stats/discrete_zipfian.rst
Comment (mdhaber) [inline doc/source/tutorial/stats/discrete_zipfian.rst:7]: Followed the language of the Zipf distribution.

File: doc/source/tutorial/stats/discrete_zipfian.rst
Comment (mdhaber) [inline doc/source/tutorial/stats/discrete_zipfian.rst:8]: Chose these to match the parameter names in Wikipedia.

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/_discrete_distns.py:1027]: Suggestions for more specific language than "special case"?

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/_discrete_distns.py:1037]: Shall I add an example that shows the equivalence as `N -> oo`?

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/_discrete_distns.py:1101]: Shall I put this on one line? I was just matching zipf.

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/_discrete_distns.py:1085]: More appropriate language than "converges"?

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/_discrete_distns.py:1077]: Used `a` to match `zipf`.

File: scipy/stats/tests/test_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/tests/test_discrete_distns.py:109]: I can add tests to test against some other reference.

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/_discrete_distns.py:1068]: Add see also?

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/_discrete_distns.py:1127]:
```suggestion
              - 3*Hna1**4) / mu2n**2 - 3
```
Why is subtracting 3 needed? I see this in `dlaplace_gen` and `betabinom_gen`, and that's exactly what was needed here.

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/_discrete_distns.py:1110]: Does this `sf` have the potential to be more precise than `1-cdf`? Here and in `_stats` below, implementation follows the reference directly; suggestions for improvement to precision welcome.

File: scipy/stats/_discrete_distns.py
Comment (ev-br) [reply scipy/stats/_discrete_distns.py:1085]: Maybe "reduces to"?

File: scipy/stats/_discrete_distns.py
Comment (ev-br) [inline scipy/stats/_discrete_distns.py:1098]: Asarrays not needed? IIRC all is converted to arrays when _argcheck is called

File: scipy/stats/_discrete_distns.py
Comment (ev-br) [inline scipy/stats/_discrete_distns.py:1108]: Being paranoid: can you check this does not modify input inplace? Esp if k is already an array

Comment (mdhaber) [reply scipy/stats/_discrete_distns.py:1108]: Yes I forgot about that for arrays. I'll change it.

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [reply scipy/stats/_discrete_distns.py:1098]: Yes I meant to check. Thanks for the reminder.

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [reply scipy/stats/_discrete_distns.py:1085]: Perfect; I'll change today.

File: scipy/stats/tests/test_discrete_distns.py
Comment (mdhaber) [reply scipy/stats/tests/test_discrete_distns.py:109]: zipfUC from R's VGAM package seems innacurate for low `a`, `N`. Trying this at [rdrr.io](https://rdrr.io/cran/VGAM/man/zipfUC.html):
```
library(VGAM)
print(dzipf(0:21, shape = 1.01, N = 20))
print(sum(dzipf(0:21, shape = 1.01, N = 20)))
print(pzipf(20, shape = 1.01, N = 20))
```
The first line prints the pmf, the second is the sum of the pmf, and the third is the cdf. The sum of the pmf values is not 1.

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [reply scipy/stats/_discrete_distns.py:1098]: Apparently not:
```
return (a >= 0) & (n == n.astype(int))
E   AttributeError: 'int' object has no attribute 'astype'
```

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/_discrete_distns.py:1073]: How should I implement this for a <=1 ? I was thinking a naive (just summing the harmonic series) in a ufunc. I could try to do something fancy from within Python that saves computation (e.g. for each distinct `a`, generate the harmonic series up to the maximum of the `n`s, perform a cumulative sum for the `n`s that are needed) but I'm not sure it's worth the trouble.

**Comment by mdhaber** [comment]
I anticipate two test failures.  Both are ultimately caused by trouble with `expect`:
```
import numpy as np
from scipy.stats import zipf, zipfian
a, n = 3.5, 10
print(zipfian.moment(0, a, n))
print(zipfian.expect(lambda x: 1, (a, n)))
```
These are supposed to both be 1, but we get:
```
1.0
1.0009914161108657
```

The second is from difficulty calculating the entropy:
```
print(zipfian.entropy(a, n))
```
and we get a warning:
```
RuntimeWarning: expect(): sum did not converge
```
Note that the existing `zipf` distribution suffers from the same problem, it's just not exposed because the "sane parameter" was chosen as `a=6.5`, for which neither distribution has any trouble. There is a formula for the entropy on Wikipedia, but it is not very stable.

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [reply scipy/stats/_discrete_distns.py]: This happens in the `moment` method. I think it is being fixed in gh-12197.

File: scipy/stats/tests/test_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/tests/test_discrete_distns.py:143]: I should remove this test against R because the R implementation is wrong. For example, run this code at [rdrr.io](https://rdrr.io/cran/VGAM/man/zipfUC.html).
```R
library(VGAM)
options(digits=16)
k <- 3
a <- 1.2589254117941673
n <- 12
pmf <- dzipf(k, N = n, shape = a)
cdf <- pzipf(k, N = n, shape = a)
print(pmf)
print(cdf)
```
For `n <= 12`, the two implementations match to machine precision.
Increase `n` to 13, and the error is ~0.001. You can tell R is wrong because if you set:
```R
k <- 0:14
```
and sum the pmf values, they don't add to 1.

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/_discrete_distns.py:1145]: I had a check for n being an integer in here, but I removed it because I don't see it for any other discrete distribution which is supposed to have an integer shape. It generalizes in a reasonable way for a > 1, but not for a < 1. Should I add in the integer check?

File: doc/source/tutorial/stats/discrete_zipfian.rst
Comment (mdhaber) [inline doc/source/tutorial/stats/discrete_zipfian.rst:22]:
```suggestion
is the :math:`N`:sup:`th` generalized harmonic number of order
```

File: doc/source/tutorial/stats/discrete_zipfian.rst
Comment (mdhaber) [inline doc/source/tutorial/stats/discrete_zipfian.rst:14]:
```suggestion
    \begin{eqnarray*} p\left(k; s, N \right) & = & \frac{1}{H_{N, s}k^{s}}\quad k \in \{1, 2, \dots, n-1, n\} \end{eqnarray*}
```

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/_discrete_distns.py:1159]:
```suggestion
```

File: scipy/stats/tests/test_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/tests/test_discrete_distns.py:170]: I think this is a fine test if I adjust the values of a it tests to be logarithmically distributed between some small number and some large number.

File: scipy/stats/tests/test_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/tests/test_discrete_distns.py:171]:
```suggestion
        # test against bare-bones implementation of Zipfian
```

File: scipy/stats/_discrete_distns.py
Comment (mckib2) [reply scipy/stats/_discrete_distns.py:1145]: I think an integer check makes sense if something breaks when `n` isn't an integer.  If it still "works" if `n` isn't an integer, then if it seems like it would be consistent with the existing stats distributions to not check I would follow precedence

File: scipy/stats/tests/test_discrete_distns.py
Comment (mckib2) [reply scipy/stats/tests/test_discrete_distns.py:143]: So I assume the goal of this test is to provide reasonable assurance that this SciPy implementation is doing a reasonable thing compared to another independent implementation, in this case R.  Is Wolfram's `ZipfDistribution` the correct distribution? If not, might be okay to omit this test if we're testing against bad values

FYI, appears that R [is using](https://github.com/cran/VGAM/blob/ad275cd8262668ccf792601c74c008f5014285d9/R/family.aunivariate.R#L4815) what you call below the "naive" method

Comment (mdhaber) [reply scipy/stats/tests/test_discrete_distns.py:143]: Yes, the idea is just to compare against an existing implementation.

>  Is Wolfram's ZipfDistribution the correct distribution?

No, that's the zeta distribution. The confusion between the two distributions (gh-11523) is why I created this PR in the first place, actually : )  So another thing I was thinking we might want to do is doc-deprecate `zipf` and give it an alias `zeta`.

Comment (mckib2) [reply scipy/stats/tests/test_discrete_distns.py:143]: Nope, I don't think it's license compatible, it's GPL 2 or 3

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [reply scipy/stats/_discrete_distns.py:1145]: OK adding integer check back.

File: scipy/stats/tests/test_discrete_distns.py
Comment (mdhaber) [reply scipy/stats/tests/test_discrete_distns.py:171]: Expanded these tests.

**Comment by rlucas7** [comment]
The 3 comes from a kurtosis of a normal/Gaussian. This is sometimes called kurtosis excess, the excess w.r.t. a Gaussian. 

For better or worse, many things in statistics are relative to a Gaussian. 

-L

**Comment by rlucas7** [comment]
> On Dec 28, 2020, at 2:13 AM, Matt Haberland <notifications@github.com> wrote:
> 
> ﻿
> @mdhaber commented on this pull request.
> 
> In scipy/stats/_discrete_distns.py:
> 
> > +    .. [1] "Zipf's Law", Wikipedia, https://en.wikipedia.org/wiki/Zipf's_law
> +    .. [2] Larry Leemis, "Zipf Distribution", Univariate Distribution
> +           Relationships. http://www.math.wm.edu/~leemis/chart/UDR/PDFs/Zipf.pdf
> +
> +    %(example)s
> +
> +    Confirm that `zipfian` reduces to `zipf` for large `n`, `a > 1`.
> +
> +    >>> from scipy.stats import zipf
> +    >>> k = np.arange(11)
> +    >>> np.allclose(zipfian.pmf(k, a=3.5, n=10000000), zipf.pmf(k, a=3.5))
> +    True
> +
> +    """
> +    def _argcheck(self, a, n):
> +        return (a >= 0) & (n > 0)
> I had a check for n being an integer in here, but I removed it because I don't see it for any other discrete distribution which is supposed to have an integer shape. It generalizes in a reasonable way for a > 1, but not for a < 1. Should I add in the integer check?
> 
I see why that’s a reasonable thing to do but for consistency with the rest of codebase I’d omit, unless people have strong opinions to include.

**Comment by rlucas7** [comment]
@mdhaber this looks mostly ready, is there anything else you wanted to change? 

If not I can take a closer look over the next few days. 

Sincerely, 

-Lucas Roberts

File: scipy/stats/_discrete_distns.py
Comment (rlucas7) [reply scipy/stats/_discrete_distns.py]: @mdhaber should this PR wait for gh-12197 to be merged then?

File: scipy/stats/tests/test_discrete_distns.py
Comment (rlucas7) [inline scipy/stats/tests/test_discrete_distns.py]: should this test be marked slow? I'm assuming it will be.

**Review by rlucas7 — COMMENTED** [review]
@mdhaber I left two (very minor) comments and it looks like there might be 1 or two open changes from @ev-br review. 

EDIT: Previous rejection sampler comment is removed, I mis-understood the distinction between *zipf* and *zeta* and here. 

I still think a rejection sampler can work here using the `zipf` approach with an additional rejection step if the drawn random variate is larger than `n`, the truncation point but I'll want to look a little closer to confirm so we can omit from this PR unless you feel strongly to include.

File: scipy/stats/_discrete_distns.py
Comment (rlucas7) [inline scipy/stats/_discrete_distns.py:1149]: Ah I see here the distinction. The rejection sampler I wrote in comment is for `zipf` not `zipfian`-I'll edit.

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [reply scipy/stats/_discrete_distns.py]: I don't think so. It's not really hurting anything to ensure it's an array.

File: scipy/stats/_discrete_distns.py
Comment (mdhaber) [inline scipy/stats/_discrete_distns.py]:
```suggestion
        k = k + 1  # # to match SciPy convention
```

**Comment by mdhaber** [comment]
> @mdhaber I left two (very minor) comments and it looks like there might be 1 or two open changes from @ev-br review.

It looks like you commented on an outdated commit. I believe @ev-br's review has been addressed, but I don't click resolve on reviewer comments. When I review PRs, I prefer to to be the one to click resolve so I can keep track of what changed in response and verify that the change sufficiently addresses the comment.

> so we can omit from this PR unless you feel strongly to include.

I'd prefer to omit. That can be an enhancement.

> I see why that’s a reasonable thing to do but for consistency with the rest of codebase I’d omit, unless people have strong opinions to include.  EDIT: I see from the conversion (now that I'm logged in on laptop) that something broke and keeping int check makes sense to fail fast.

Well, I don't know if there's an error for non-integers, but when `a < 1` we sum `n` terms of a sequence, and I'm not sure how to sum a fraction of a term, so the result would be wrong.

File: scipy/stats/tests/test_discrete_distns.py
Comment (mdhaber) [reply scipy/stats/tests/test_discrete_distns.py]: In the version of the code you commented on, `test_zipfian` is equivalent to the test now called `test_zipfian_asymptotic`. 

For `a` greater than 1, the generalized harmonic numbers are not calculated by taking the sum of a series; they use the zeta function, so the magnitude of `n` doesn't really affect the speed. The test is for `k = np.arange(1, 21)`, and it doesn't take too long to do 21 of these calculations.

This test is slow:
```
7.94s call     stats/tests/test_discrete_basic.py::test_discrete_basic[zipfian-arg17-True]
```
so I should check skip whatever is taking a long time there - maybe `rvs`.

Comment (rlucas7) [reply scipy/stats/tests/test_discrete_distns.py]: these are comments on outdated diff, marking resolved

**Review by rlucas7 — APPROVED** [review]

**Comment by rlucas7** [comment]
> > @mdhaber I left two (very minor) comments and it looks like there might be 1 or two open changes from @ev-br review.
> 
> It looks like you commented on an outdated commit. 

Yeah, sometimes github doesn't show me the latest commit diffs, I'm probably doing something wrong. 

>I believe @ev-br's review has been addressed, but I don't click resolve on reviewer comments. When I review PRs, I prefer to to be the one to click resolve so I can keep track of what changed in response and verify that the change sufficiently addresses the comment.

Yeah I agree, comment does look addressed, maybe @ev-br can comment/close that one? 

> > so we can omit from this PR unless you feel strongly to include.
> 
> I'd prefer to omit. That can be an enhancement.

That's fine. I can open a PR once this is merged, it's pretty clear to me the change and cases.
 
> > I see why that’s a reasonable thing to do but for consistency with the rest of codebase I’d omit, unless people have strong opinions to include.  EDIT: I see from the conversion (now that I'm logged in on laptop) that something broke and keeping int check makes sense to fail fast.
> 
> Well, I don't know if there's an error for non-integers, but when `a < 1` we sum `n` terms of a sequence, and I'm not sure how to sum a fraction of a term, so the result would be wrong.

Right, `n` should be int here, otherwise things won't be correct.

AFAICS the PR looks ready to merge, let's give @ev-br a week or so to comment. It's easy to miss messages on github though so if no comments for changes within a week or so feel free to ping me and I can merge this one.

File: scipy/stats/tests/test_discrete_distns.py
Comment (mdhaber) [reply scipy/stats/tests/test_discrete_distns.py]: For now I've effectively marked the `zipfian` `test_discrete_basic` as slow, so it only runs when test mode is full.

**Comment by mdhaber** [comment]
Thanks very much @rlucas7! If you could check one thing before then - I just pushed a commit that marks certain cases of `test_discrete_basic` as slow, so they only run when test mode is full.  This is the same way it is done in `test_continuous_basic.py`. I went ahead and marked both `zipfian` and `nhypergeom`, since that was pretty slow, too.

File: scipy/stats/tests/test_discrete_basic.py
Comment (rlucas7) [inline scipy/stats/tests/test_discrete_basic.py:25]: I was wondering how the tested get marked for pytest to still pickup the slow marker, it's in this line. LGTM.

**Review by rlucas7 — APPROVED** [review]
@mdhaber LGTM. 
I'll leave it open until this Sunday Jan 31st, if no further comments by morning EST I'll merge.

**Comment by rlucas7** [comment]
Everything green and no further comments so squashing and merging. 
Thanks @mdhaber and @ev-br for the review. 

@mdhaber I'll add an entry to the release notes wiki-please update if anything needs adding/correcting in the release notes entry

**Comment by mdhaber** [comment]
Thanks @rlucas7 @ev-br!
The original issue noted confusion about the `zipf` distribution, which could mean `zipfian` but is also known as `zeta`. There is less chance of confusion between `zeta` and `zipfian`. Should I suggest on the mailing list that we add `zeta` as an alias for the existing `zipf` and deprecate `zipf`?

**Comment by rlucas7** [comment]
@mdhaber

I think that might be a good idea but I’m not sure what others think, I was confused by the difference initially but not necessarily because of naming here in scipy. Rather that the naming seems to be inconsistent acrosss papers and resources I have. 

Mailing list would be good to get others thoughts so we come up with better naming-naming is difficult.

-L

**Comment by mdhaber** [comment]
Thanks. I actually don't think I'll bring it up. I still think it's a good idea, there are just higher priority things (even if this would be easy/fast).

**Pull request merged**
